python/GafferUI/ContextQueryUI.py

```diff
@@ -0,0 +1,408 @@
+			{ plug["exists"] for plug in self.getPlugs() },
+			horizontalAlignment = GafferUI.Label.HorizontalAlignment.Right
+		)
+		existsLabelWidget.label()._qtWidget().setFixedWidth( 40 )
+		self.__row.append(
+			existsLabelWidget,
+			verticalAlignment = GafferUI.Label.VerticalAlignment.Top
+		)
+
+		self.__row.append(
+			GafferUI.BoolPlugValueWidget( { plug["exists"] for plug in self.getPlugs() } ),
+			verticalAlignment = GafferUI.Label.VerticalAlignment.Top
+		)
+
+		valueLabelWidget = GafferUI.LabelPlugValueWidget(
+			{ plug["value"] for plug in self.getPlugs() },
+			horizontalAlignment = GafferUI.Label.HorizontalAlignment.Right
+		)
+		valueLabelWidget.label()._qtWidget().setFixedWidth( 40 )
+		self.__row.append(
+			valueLabelWidget,
+			verticalAlignment = GafferUI.Label.VerticalAlignment.Top
+		)
+
+		self.__row.append(
+			GafferUI.PlugValueWidget.create( { plug["value"] for plug in self.getPlugs() } )
+		)
+
+	def setPlugs( self, plugs ) :
+
+		GafferUI.PlugValueWidget.setPlugs( self, plugs )
+
+		self.__row[0].setPlugs( plugs )
+		self.__row[1].setPlugs( { plug["exists"] for plug in plugs } )
+		self.__row[2].setPlugs( { plug["exists"] for plug in plugs } )
+		self.__row[3].setPlugs( { plug["value"] for plug in plugs } )
+		self.__row[4].setPlugs( { plug["value"] for plug in plugs } )
+
+	def hasLabel( self ) :
+
+		return True
+
+	def childPlugValueWidget( self, childPlug ) :
+
+		for w in self.__row :
+			if childPlug in w.getPlugs() :
+				return w
+
+		return None
+
+	def setReadOnly( self, readOnly ) :
+
+		if readOnly == self.getReadOnly() :
+			return
+
+		GafferUI.PlugValueWidget.setReadOnly( self, readOnly )
+
+		for w in self.__row :
+			w.setReadOnly( readOnly )
+
+
+##########################################################################
+# Metadata
+##########################################################################
+
+Gaffer.Metadata.registerNode(
+
+	Gaffer.ContextQuery,
+
+	"description",
+	"""
+	Queries variables from the current context, creating outputs for each variable.
+	""",
+
+	plugs = {
+
+		"queries" : [
+
+			"description",
+			"""
+			The context variables to be queried - arbitrary numbers of context
+			variables may be added as children of this plug via the user interface,
+			or via python. Each child is a `NameValuePlug` whose `name` plug is
+			the context variable to query, and whose `value` plug is the default
+			value to use if the variable does not exist in the context with an
+			appropriate type.
+			""",
+
+			"plugValueWidget:type", "GafferUI.LayoutPlugValueWidget",
+
+			"layout:customWidget:footer:widgetType", "GafferUI.ContextQueryUI._ContextQueryFooter",
+			"layout:customWidget:footer:index", -1,
+
+			"nodule:type", "",
+
+		],
+
+		"queries.*" : [
+
+			"description",
+			"""
+			The default value to output if the variable does not exist.
+			""",
+
+			"spreadsheet:columnName", functools.partial( __getLabel, parentPlug = "queries", sanitise = True ),
```

Comment:
I don't know what we're hoping to achieve here - I guess we're trying to get the spreadsheet column to be named after `query["name"].getValue()`, in a similar way to what we do for ArnoldAttributes etc? In any case, it's creating this error when trying to create a spreadsheet (via the "Value and Switch") option :

```
  File "/home/john/dev/build/gaffer62Python3/python/GafferUI/SpreadsheetUI/_Algo.py", line 56, in createSpreadsheet
    addColumn( spreadsheet, plug )
  File "/home/john/dev/build/gaffer62Python3/python/GafferUI/SpreadsheetUI/_Algo.py", line 88, in addColumn
    columnName = Gaffer.Metadata.value( plug, "spreadsheet:columnName" ) or plug.getName()
  File "/home/john/dev/build/gaffer62Python3/python/GafferUI/ContextQueryUI.py", line 56, in __getLabel
    qPlug = n.queryPlugFromOutPlug( plug )
IECore.Exception: ContextQuery : Plug not in hierarchy.
```

I don't think we're likely to drive this with a spreadsheet, so rather than worry about it, I think we'd be better off just removing the metadata for now (same for ShaderQuery, since I assume that's where it came from).

Reply:
> I guess we're trying to get the spreadsheet column to be named after query["name"].getValue()

Yes, I had included it for `ShaderQuery` as part of the effort to make the UI a little more friendly - the partial function is called for tooltips as well.

Strangely it's working for adding the (not terribly useful) `name` plug to a `Spreadsheet`, but not `value`. If it's worth moving the column label handling to `queries.*.value` metadata, I'm happy sorting out the bug as part of `ShaderQuery`.

Reply:
I'm still not sure why we'd be adding these to a spreadsheet, or what would be useful in that case :

- If we're adding the whole NameValuePlug (the case that errors now), then it's not meaningful to name the column after `query["name"].getValue()`, because the name will be dynamic (driven by the spreadsheet).
- If we're adding just the `query["name"]` to the spreadsheet, the same applies.
- I can't think of a plausible reason I'd add `query["value"]` to a spreadsheet - I don't know why I'd want a dynamic default value.

Unless we have a good use case, I'm inclined to save ourselves the bother and just remove the spreadsheet metadata.